// zos/lib/models/stdlib/Stdlib.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _zosLib = require('zos-lib');

var _semver = require('semver');

var _semver2 = _interopRequireDefault(_semver);

var _StdlibProvider = require('./StdlibProvider');

var _StdlibProvider2 = _interopRequireDefault(_StdlibProvider);

var _StdlibDeployer = require('./StdlibDeployer');

var _StdlibDeployer2 = _interopRequireDefault(_StdlibDeployer);

var _StdlibInstaller = require('./StdlibInstaller');

var _StdlibInstaller2 = _interopRequireDefault(_StdlibInstaller);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Stdlib {
  static fetch() {
    return _StdlibProvider2.default.from(...arguments);
  }

  static async deploy() {
    return await _StdlibDeployer2.default.deploy(...arguments);
  }

  static satisfiesVersion(version, requirement) {
    return !requirement || version === requirement || _semver2.default.satisfies(version, requirement);
  }

  static validateSatisfiesVersion(version, requirement) {
    if (!Stdlib.satisfiesVersion(version, requirement)) {
      throw Error(`Required stdlib version ${requirement} does not match stdlib package version ${version}`);
    }
  }

  constructor(nameAndVersion) {
    this._parseNameVersion(nameAndVersion);
  }

  contracts() {
    return this.getPackage().contracts;
  }

  contract(alias) {
    return this.contracts()[alias];
  }

  hasContract(alias) {
    if (!this.contracts()) return false;
    return !_lodash2.default.isEmpty(this.contract(alias));
  }

  async install() {
    await _StdlibInstaller2.default.call(this.nameAndVersion);
  }

  getPackage() {
    if (this._packageJson) return this._packageJson;
    const filename = `node_modules/${this.name}/zos.json`;
    this._packageJson = _zosLib.FileSystem.parseJson(filename);
    return this._packageJson;
  }

  _parseNameVersion(nameAndVersion) {
    const [name, version] = nameAndVersion.split('@');
    this.name = name;
    this.nameAndVersion = nameAndVersion;

    const packageVersion = this.getPackage().version;
    Stdlib.validateSatisfiesVersion(packageVersion, version);

    this.version = version || tryWithCaret(packageVersion);
  }
}

exports.default = Stdlib;
function tryWithCaret(version) {
  const cleaned = _semver2.default.clean(version);
  return cleaned ? `^${cleaned}` : version;
}
